### PR TITLE
Update breaktimer from 0.7.2 to 0.7.3

### PR DESCRIPTION
**Pull request by vladimyr**
Add binary, `uninstall` stanza & update `zap` stanza.
Continuation of #80775 

After making all changes to the cask:

- [x] `brew cask audit --download {{cask_file}}` is error-free.
- [x] `brew cask style --fix {{cask_file}}` reports no offenses.
- [x] The commit message includes the cask’s name and version.

### CODE REVIEW

File: Casks/breaktimer.rb [hunk 1]
```diff
@@ -1,17 +1,28 @@
...
   # github.com/tom-james-watson/breaktimer-app was verified as official when first introduced to the cask
   url "https://github.com/tom-james-watson/breaktimer-app/releases/download/v#{version}/BreakTimer.dmg"
   appcast 'https://github.com/tom-james-watson/breaktimer-app/releases.atom'
   name 'BreakTimer'
   homepage 'https://breaktimer.app/'
 
+  auto_updates true
+
   app 'BreakTimer.app'
+  binary "#{appdir}/BreakTimer.app/Contents/MacOS/BreakTimer", target: 'breaktimer'
```
Comment (vitorgalvao) [inline Casks/breaktimer.rb:14]: Is this linking recommended anywhere, by upstream?

Comment (vladimyr) [reply Casks/breaktimer.rb:14]: Yes, this refers to Linux version but is also applicable to macOS one https://github.com/tom-james-watson/breaktimer-app/blob/master/README.md#is-there-a-way-to-control-the-app-via-the-commmand-line

**Comment by SMillerDev** [comment]
How did you find all these extra parameters by the way? Brew audit should probably find those.

**Comment by vitorgalvao** [comment]
@SMillerDev It’s hard / impractical for `audit` alone to find all of those, because several require the app to have run. For example, unless the vendor has an online list of all files created by the application—which virtually none do—filling `zap` mostly needs to be done by investigation after using the app.

That said, CI can (and does) find some of these.

Some examples, based on this PR:

#### `auto_updates true`

Could be implied if [`find-appcast`](https://github.com/Homebrew/homebrew-cask/blob/master/developer/bin/find-appcast) is able to find something, though sometimes it can only find [part of something, because we don’t yet know all forms an electron-builder appcast may take](https://github.com/Homebrew/homebrew-cask/blob/master/developer/bin/find-appcast#L74L76).

`auto_updates true` always implies an [`appcast`](https://github.com/Homebrew/homebrew-cask/blob/master/doc/cask_language_reference/stanzas/appcast.md), but `find-appcast` is not able to find them all because there’s no method that works every time. Several appcasts are only discovered by inspecting network traffic and telling the app to try auto-updating.

Conversely, `appcast` does not necessarily imply `auto_updates`.

In theory `find-appcast` could find an `appcast` that isn’t used for auto-updating, though in practice I doubt that’d happen. An example on a popular app that internally checks for updates but does not update itself (it only informs you there’s an update available and sends you to the website) is [Calibre](https://calibre-ebook.com/).

`appcast`, while highly desirable, is not on the PR template partly because it may require technical knowledge to figure out. I usually check for it on new submissions, as does @vladimyr.

#### `uninstall quit`

That one is far from mandatory. In most cases it’s a nice to have but little more. It could be deduced by inspecting the `CFBundleIdentifier` of the app’s `Info.plist` (easy enough), though a better use of time would be to implement the [`bundle_id`](https://github.com/Homebrew/homebrew-cask/issues/16339) stanza, which would help here and in the `zap` example.

This one is detected by the CI, but only on `pkg`s, in cases those launch an app.

With Catalina and beyond, that check [may hinder](https://github.com/Homebrew/homebrew-cask/issues/70521), rather than help.

#### `uninstall launchctl`

It’s up to an app—and often its users—if they install one of these or not, and I know of no good way to detect which an app does install other than by deducing them yourself or by comparing which `launchd` jobs you had before installing the application with the ones you have after.

That becomes a problem for `audit` to check, because it requires a pristine environment and installing the app with a before and after, which most users won’t have.

Again, this one is detected by CI, but only for `pkg`s because those tend to install everything at once. Doing this for `app` checks would require opening the app inside CI, which might still not detect the majority of cases and might be problematic anyway with Catalina and beyond.

#### More `uninstall` notes

You can see more on the [`uninstall` documentation](https://github.com/Homebrew/homebrew-cask/blob/master/doc/cask_language_reference/stanzas/uninstall.md#uninstall-key-launchctl) and the [`developer/bin` scripts](https://github.com/Homebrew/homebrew-cask/tree/master/developer/bin).

#### On CI

We still have some CI problems (false negatives) that need investigating, and those would be helpful to fix. Specifically:

* A `pkg` is installed fine on user’s machines but CI always fails (e.g. [`parallels`](https://github.com/Homebrew/homebrew-cask/blob/master/Casks/parallels.rb)).
* `uninstall` works fine on user’s machines, but CI still says things need to be removed and asks to add those to the cask (like `launchctl:`), despite them being present in the cask and working.

**Comment by SMillerDev** [comment]
That's a lot of information on why you think it's not possible. But how can you find the then? Is really the only way to submit a pull request and wait for someone else, that knows how, to contribute these?

**Comment by vitorgalvao** [comment]
> That's a lot of information on why you think it's not possible.

My exact words:

> It’s hard / impractical for `audit` alone to find all of those

“Hard” and “impractical” do not mean impossible. Also, I made it clear that while it’s hard for `audit` to do *on it’s own*, I’ve specifically pointed out the cases where CI can do it *and how it does it*.

I gave you a lot of information because we have a lot of information. There’s no point in letting you bump into the same problems and frustrations we have. If you want to tackle this, you should be doing it with the knowledge we’ve already accumulated, not from scratch. It’s not like we enjoy doing things manually; we’ve already automated plenty and stopped when we got stuck.

> But how can you find the then?

It’s all in the documentation that I linked. It’s comprehensive and I’ll be happy to fill any gaps.

App developers are incredibly inconsistent (@core-code and @suschizu would be able to confirm, they’ve seen more true cases from the wild than most people), which is why so many things require workarounds or manual inspection.

> Is really the only way to submit a pull request and wait for someone else, that knows how, to contribute these?

To our knowledge, no one has ever come up with a way to programatically guess all files an application creates (`zap`). Especially for closed-source applications, I’m guessing it would take quite a lot of effort to build a general solution. Even specialised applications (app zappers) aren’t perfect. But we could get *close* with some methods, like the `bundle_id` stanza mentioned above.

Please take my words as I write them. My previous post was nothing but assistance towards you. If you find ways to get us unstuck and improve our process—**and nowhere have I claimed it’s at the best it can be**—that would be great, but you shouldn’t have to relearn what we already know.

You’re free to either accept or refuse my help, but I will respectfully request that when I offer you my help, you read it for what it is: information to give you a head-start. It took me a while to write all that. If I had any desire to block your progress—*which I most definitely don’t and it would be counter-productive to myself*—I wouldn’t have bothered. But I did, because I want you to succeed in making our work easier and more accurate.

**Comment by SMillerDev** [comment]
I get that some parts are hard to detect but I would guess you're not installing a new macOS VM to diff what files were created every time you review a pull request. I don't need to have all the information you ever amassed on a subject, if anything that's highly demotivating and makes it difficult to find the answer I'm looking for. I just want to know: "if you would want to reproduce the information that was added in this review, how would you do it".

**Comment by core-code** [comment]
> App developers are incredibly inconsistent 

confirmed :)


> To our knowledge, no one has ever come up with a way to programatically guess all files an application creates (zap)

i guess this is impossible for non-sandboxed apps in the general case. the only way to catch /some/ of them is to launch the app and actually see which ones it creates. which i think is similar to what Travis does in some cases. @ran-dall i think you mentioned the ability of your ML system to automate some of this?

> if anything that's highly demotivating and makes it difficult to find the answer I'm looking for. 

ugh. i cannot imagine why one would be demotivated by a kind and comprehensive offer to help. 

on the other hand, i can easily imagine that it would be 'demotivating' to receive replies like these after taking so much time trying to be helpful.

**Comment by SMillerDev** [comment]
> ugh. i cannot imagine why one would be demotivated by a kind and comprehensive offer to help.

That might just be my personal quirks, but if I ask a question that is meant to be relatively easy and someone responds with 2 pages full of documents and links I'm more inclined to just forget about the question then to dig through all of it to find the answer I was looking for. Especially if I see at a glance that it features a couple of suggestions for other work I could do.

> on the other hand, i can easily imagine that it would be 'demotivating' to receive replies like these after taking so much time trying to be helpful.

The effort that went into the reply is very much appreciated, no matter if I like the outcome of it or not. All Homebrew maintainers are doing a great job in keeping the project going and any effort they put in is definitely appreciated.

**Comment by core-code** [comment]
> All Homebrew maintainers are doing a great job in keeping the project going and any effort they put in is definitely appreciated.

fully agreed. lets keep up the good work.

**Comment by ran-dall** [comment]
> @ran-dall i think you mentioned the ability of your ML system to automate some of this?

The ML does have ways of making this happen yes, but as @vitorgalvao said, it's far for detecting everything perfectly; I'd say it's about 80% accurate though. There's also @nrlquaker's project which does a decent job; https://github.com/nrlquaker/homebrew-createzap, but yet again not perfect.

**Comment by SMillerDev** [comment]
I think an important thing I'm seeing here is that we shouldn't try and wait for a perfect system if we can already shave of some hours of work every week (just guessing) with a system that isn't completely terrible.

**Comment by vitorgalvao** [comment]
> if we can already shave of some hours of work every week (just guessing)

I don’t think any maintainer is spending any time specifically on `zap`s. The stanza is nice to have but also completely optional and non-crucial.

In that sense, making it part of `audit` would make more work, not less.

But an optional system that helps one generate zaps should have positive results.

**Comment by SMillerDev** [comment]
> I don’t think any maintainer is spending any time specifically on zaps. The stanza is nice to have but also completely optional and non-crucial.

Someone spend time fixing mine :shrug:

**Comment by vitorgalvao** [comment]
> Someone spend time fixing mine

This was not a fix, it was an addition. That’s how `zap` works, you add things *as you find they exist*.

> If it's caught by an audit it's up to the submitter instead of the maintainers though

How would you have designed an `audit` for this? From the five added lines, four depend on knowing the app’s bundle ID, two of which would not be accurate even with it (they have a suffix), and one of those uses a glob.

Is the idea to have an `audit` that makes up files in common locations and tells users to add them to `zap`, in the hopes they exist? Might as well just edit the core to auto-`zap` common patterns—less maintenance and leaner casks. We’ll also get into the same problem of every app zapper, in that we may be wrong and delete the wrong things.

**Comment by SMillerDev** [comment]
> I frequently fix PRs in place to get things moving faster and so we don’t get swamped by PRs in need of changes.

I think that might be easier to fix with stalebot. If people can rely on you to fix their shit they won't do it for themselves.

> Is the idea to have an audit that makes up files in common locations and tells users to add them to zap, in the hopes they exist?

I was thinking of checking common locations for files that match a specific pattern (including the token for example) and if they're found you can tell people to add it. If they're not found that's also fine and maybe someone else will add them.

**Comment by vitorgalvao** [comment]
> I think that might be easier to fix with stalebot.

It’s not. When I do fixes, they’re generally small and I know them offhand; they take *way* less time to do than to explain how to do. Having a bunch of PRs waiting for reply creates clutter and makes maintenance harder. I know from experience what works best for my maintenance style. Github’s addition of the feature to allow edits from maintainers has been nothing but positive.

Complex fixes that require contributor feedback are marked as such.

> If people can rely on you to fix their shit they won't do it for themselves.

Nothing could be further from the truth. I’ve been doing it this way since the feature has been available, and never have users taken advantage. If anything, they’re grateful for the corrections and are *more* careful on the next submissions.

> I was thinking of checking common locations for files that match a specific pattern (…) and if they're found you can tell people to add it.

So it only works if the user has the software installed, and has been using it, correct?

Aren‘t `audit`s limited to fail or pass? I thought you couldn’t do suggestions (which these need to be).

> (including the token for example)

That will produce few results and has a high risk of false positives. A safer way that will produce more and better results is using the bundle ID.

**Comment by ran-dall** [comment]
> Nothing could be further from the truth. I’ve been doing it this way since the feature has been available, and never have users taken advantage. If anything, they’re grateful for the corrections and are more careful on the next submissions.

I concur with this statement

> So it only works if the user has the software installed, and has been using it, correct?

Yes, with `createzap` it only works if you have the app installed and have opened it at least once.

**Pull request merged**